hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpression;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Implements evaluation of SQL comparison predicates.
+ *
+ * @see ComparisonMode
+ */
+public final class ComparisonPredicate extends BiExpression<Boolean> implements IdentifiedDataSerializable {
+
+    private ComparisonMode mode;
+
+    public ComparisonPredicate() {
+        // No-op.
+    }
+
+    private ComparisonPredicate(Expression<?> left, Expression<?> right, ComparisonMode mode) {
+        super(left, right);
+        this.mode = mode;
+    }
+
+    public static ComparisonPredicate create(Expression<?> left, Expression<?> right, ComparisonMode comparisonMode) {
+        assert left.getType().equals(right.getType());
+        return new ComparisonPredicate(left, right, comparisonMode);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_COMPARISON;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @SuppressFBWarnings(value = "NP_BOOLEAN_RETURN_NULL", justification = "Any SQL expression may return null")
+    @Override
+    public Boolean eval(Row row, ExpressionEvalContext context) {
+        Object left = operand1.eval(row, context);
+        if (left == null) {
+            return null;
+        }
+
+        Object right = operand2.eval(row, context);
+        if (right == null) {
+            return null;
+        }
+
+        Comparable leftComparable = (Comparable) left;
+        Comparable rightComparable = (Comparable) right;
+
+        int order = leftComparable.compareTo(rightComparable);
```

Comment:
Is it guaranteed that two operands are coerced to the same type by this point?

Reply:
Yes, it is.